core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -203,9 +203,24 @@ class Partition(val topic: String,
     allReplicasMap.remove(replicaId)
   }
 
-  def removeFutureLocalReplica() {
+  def futureReplicaDirChanged(newDestinationDir: String): Boolean = {
+    inReadLock(leaderIsrUpdateLock) {
+      getReplica(Request.FutureLocalReplicaId) match {
+        case Some(futureReplica) =>
+          if (futureReplica.log.get.dir.getParent != newDestinationDir)
```

Comment:
This is not the only place where we access the log of the future replica. For example, we also have `appendRecordsToFutureReplica`. Do we need the lock in those cases as well?

Reply:
Great point!

Currently it does not cause problem even though we don't have read lock in `appendRecordsToFutureReplica()`. The potential concern is that the `removeFutureLocalReplica()` is called when `appendRecordsToFutureReplica()` is called. It appears that the current implementation guarantees that this won't happen by ensuring that we always call `replicaAlterLogDirsManager.removeFetcherForPartitions(...)` before `removeFutureLocalReplica()`.

But it does take non-trivial effort to figure out whether it is OK not to have read lock. Thus it can easily break when developer updates Kafka code in the future. I have updated the code to use readLock in `appendRecordsToFutureReplica()` so that the code will be safer and easier to understand.

I have gone over the code in `Partition.scala` where future replica is accessed to ensure that 1) readLock is always used when we write records to future replica and 2) writeLock is always used when we add/remove future replica. So the logic around future replica should be safe now 😄

The previous motivation for not adding readLock here is to reduce the contention between ReplicaAlterLogDirsThread when it appends data to future replica and request handler thread when it updates ISR. Now I think about it again, it seems that the code for updating ISR is very fast so it should be OK to grab the read lock in `appendRecordsToFutureReplicaI()`.